### ISSUE TITLE
Layouts | Adding a Media Library Picker field blows up Layouts

### ISSUE THREAD

**Issue by jagraz** [OrchardCMS/Orchard#5489]
*** Repro Steps ***

1) create a new content type
2) add a media library picker field to it
3) create a new content of the content type

*** Desired results ***
Layout part content saves; can be edited

*** Actual Results ***
Page reloads; layout stage is empty. 

This also happens if you add the media library picker field to the content type after creating a layout. 

I get this error in the logs:

2015-07-08 10:27:37,799 [23] Orchard.ContentManagement.Drivers.Coordinators.ContentPartDriverCoordinator - Default - TargetInvocationException thrown from IContentPartDriver by Orchard.Layouts.Drivers.LayoutPartDriver
 http://orchard19.local.onestop.com/Admin/Contents/Create/attorney
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Orchard.MediaLibrary.Fields.MediaLibraryPickerField.get_MediaParts()
   at ASP._Page_Modules_Orchard_MediaLibrary_Views_Fields_MediaLibraryPicker_Summary_cshtml.Execute() in c:\Onestop.GIT\Orchard19\src\Orchard.Web\Modules\Orchard.MediaLibrary\Views\Fields\MediaLibraryPicker.Summary.cshtml:line 8
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.Mvc.WebViewPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at System.Web.Mvc.Html.PartialExtensions.Partial(HtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.Render(ShapeDescriptor shapeDescriptor, DisplayContext displayContext, HarvestShapeInfo harvestShapeInfo, HarvestShapeHit harvestShapeHit) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 151
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.<>c__DisplayClass26.<>c__DisplayClass28.<Discover>b__15(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 133
   at Orchard.DisplayManagement.Descriptors.ShapeAlterationBuilder.<>c__DisplayClass3.<>c__DisplayClass5.<BoundAs>b__2(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAlterationBuilder.cs:line 56
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Process(ShapeBinding shapeBinding, IShape shape, DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 203
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Execute(DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 99
   at Orchard.DisplayManagement.Implementation.DisplayHelper.ShapeExecute(Object shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 86
   at Orchard.DisplayManagement.Implementation.DisplayHelper.Invoke(String name, INamedEnumerable`1 parameters) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 44
   at Orchard.DisplayManagement.Implementation.DisplayHelper.TryInvoke(InvokeBinder binder, Object[] args, Object& result) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 29
   at CallSite.Target(Closure , CallSite , Object , Object )
   at Orchard.Core.Shapes.CoreShapes.ContentZone(Object Display, Object Shape, TextWriter Output) in c:\Onestop.GIT\Orchard19\src\Orchard.Web\Core\Shapes\CoreShapes.cs:line 286
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Orchard.DisplayManagement.Descriptors.ShapeAttributeStrategy.ShapeAttributeBindingStrategy.PerformInvoke(DisplayContext displayContext, MethodInfo methodInfo, Object serviceInstance) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAttributeStrategy\ShapeAttributeBindingStrategy.cs:line 64
   at Orchard.DisplayManagement.Descriptors.ShapeAttributeStrategy.ShapeAttributeBindingStrategy.<>c__DisplayClass4.<CreateDelegate>b__3(DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAttributeStrategy\ShapeAttributeBindingStrategy.cs:line 55
   at Orchard.DisplayManagement.Descriptors.ShapeAlterationBuilder.<>c__DisplayClass3.<>c__DisplayClass5.<BoundAs>b__2(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAlterationBuilder.cs:line 56
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Process(ShapeBinding shapeBinding, IShape shape, DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 203
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Execute(DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 99
   at Orchard.DisplayManagement.Implementation.DisplayHelper.ShapeExecute(Object shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 86
   at Orchard.DisplayManagement.Implementation.DisplayHelper.Invoke(String name, INamedEnumerable`1 parameters) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 44
   at Orchard.DisplayManagement.Implementation.DisplayHelper.TryInvoke(InvokeBinder binder, Object[] args, Object& result) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 29
   at CallSite.Target(Closure , CallSite , Object , Object )
   at ASP._Page_Modules_Orchard_Layouts_Views_ContentField_cshtml.Execute() in c:\Onestop.GIT\Orchard19\src\Orchard.Web\Modules\Orchard.Layouts\Views\ContentField.cshtml:line 8
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.Mvc.WebViewPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at System.Web.Mvc.Html.PartialExtensions.Partial(HtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.Render(ShapeDescriptor shapeDescriptor, DisplayContext displayContext, HarvestShapeInfo harvestShapeInfo, HarvestShapeHit harvestShapeHit) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 151
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.<>c__DisplayClass26.<>c__DisplayClass28.<Discover>b__15(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 133
   at Orchard.DisplayManagement.Descriptors.ShapeAlterationBuilder.<>c__DisplayClass3.<>c__DisplayClass5.<BoundAs>b__2(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAlterationBuilder.cs:line 56
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Process(ShapeBinding shapeBinding, IShape shape, DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 203
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Execute(DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 99
   at Orchard.DisplayManagement.Implementation.DisplayHelper.ShapeExecute(Object shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 86
   at Orchard.DisplayManagement.Implementation.DisplayHelper.Invoke(String name, INamedEnumerable`1 parameters) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 44
   at Orchard.DisplayManagement.Implementation.DisplayHelper.TryInvoke(InvokeBinder binder, Object[] args, Object& result) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 29
   at CallSite.Target(Closure , CallSite , Object , Object )
   at ASP._Page_Modules_Orchard_Layouts_Views_Elements_ContentField_Design_cshtml.Execute() in c:\Onestop.GIT\Orchard19\src\Orchard.Web\Modules\Orchard.Layouts\Views\Elements\ContentField.Design.cshtml:line 1
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.Mvc.WebViewPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.RenderRazorViewToString(String path, DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 174
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.Render(ShapeDescriptor shapeDescriptor, DisplayContext displayContext, HarvestShapeInfo harvestShapeInfo, HarvestShapeHit harvestShapeHit) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 157
   at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.<>c__DisplayClass26.<>c__DisplayClass28.<Discover>b__15(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 133
   at Orchard.DisplayManagement.Descriptors.ShapeAlterationBuilder.<>c__DisplayClass3.<>c__DisplayClass5.<BoundAs>b__2(DisplayContext displayContext) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Descriptors\ShapeAlterationBuilder.cs:line 56
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Process(ShapeBinding shapeBinding, IShape shape, DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 203
   at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Execute(DisplayContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 99
   at Orchard.DisplayManagement.Implementation.DisplayHelper.ShapeExecute(Object shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 86
   at Orchard.DisplayManagement.ShapeDisplay.Display(Object shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\ShapeDisplay.cs:line 33
   at Orchard.DisplayManagement.ShapeDisplay.Display(Shape shape) in c:\Onestop.GIT\Orchard19\src\Orchard\DisplayManagement\ShapeDisplay.cs:line 20
   at Orchard.Layouts.Services.LayoutEditorFactory.RenderElement(Element element, DescribeElementsContext describeContext, String displayType)
   at Orchard.Layouts.Services.LayoutEditorFactory.<>c__DisplayClassa.<GetConfigurationData>b__9(ElementDescriptor descriptor)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.Linq.JToken.FromObjectInternal(Object o, JsonSerializer jsonSerializer)
   at Orchard.Layouts.Services.LayoutEditorFactory.GetConfigurationData(IContent content)
   at Orchard.Layouts.Services.LayoutEditorFactory.Create(String layoutData, String sessionKey, Nullable`1 templateId, IContent content)
   at Orchard.Layouts.Services.LayoutEditorFactory.Create(LayoutPart layoutPart)
   at Orchard.Layouts.Drivers.LayoutPartDriver.<>c__DisplayClass14.<Editor>b__13()
   at Orchard.ContentManagement.Drivers.ContentPartDriver`1.<>c__DisplayClass1b.<ContentShape>b__1a(BuildShapeContext ctx) in c:\Onestop.GIT\Orchard19\src\Orchard\ContentManagement\Drivers\ContentPartDriver.cs:line 217
   at Orchard.ContentManagement.Drivers.ContentPartDriver`1.<>c__DisplayClass21.<ContentShapeImplementation>b__20(BuildShapeContext ctx) in c:\Onestop.GIT\Orchard19\src\Orchard\ContentManagement\Drivers\ContentPartDriver.cs:line 226
   at Orchard.ContentManagement.Drivers.ContentShapeResult.ApplyImplementation(BuildShapeContext context, String displayType) in c:\Onestop.GIT\Orchard19\src\Orchard\ContentManagement\Drivers\ContentShapeResult.cs:line 45
   at Orchard.ContentManagement.Drivers.ContentShapeResult.Apply(BuildEditorContext context) in c:\Onestop.GIT\Orchard19\src\Orchard\ContentManagement\Drivers\ContentShapeResult.cs:line 25
   at Orchard.ContentManagement.Drivers.Coordinators.ContentPartDriverCoordinator.<>c__DisplayClass10.<UpdateEditor>b__f(IContentPartDriver driver) in c:\Onestop.GIT\Orchard19\src\Orchard\ContentManagement\Drivers\Coordinators\ContentPartDriverCoordinator.cs:line 65
   at Orchard.InvokeExtensions.Invoke[TEvents](IEnumerable`1 events, Action`1 dispatch, ILogger logger) in c:\Onestop.GIT\Orchard19\src\Orchard\InvokeExtensions.cs:line 17

**Comment by sfmskywalker** [comment]
The issue is that the layout editor tries and renders the media library picker field for new content items that have this field attached. However, the `MediaLibraryPickerField.MediaParts` property's `_contentItems` lazy field is only initialized when loading an existing content item (it's initialized during the Loaded event of the content item).

The simplest fix is to check for null in the `MediaParts` property getter.
The better fix I think is to replace the Lazy<T> with a LazyField<T> and instantiate this field inside of the `MediaLibraryPickerField` class (same approach as with other fields and parts). That way we'll always have an actual instance of the lazy field, even if it's not yet initialized, which is fine.

**Comment by sfmskywalker** [comment]
I decided to go with the simplest solution, because even if I refactored `Lazy<T>` with `LazyField<T>`, the value of the lazy field would still return null, so I'd have to return an empty collection anyway, so might as well check for null against the internal lazy field like so:

```C#
public IEnumerable<MediaPart> MediaParts { 
    get {
        return _contentItems != null ? _contentItems.Value : Enumerable.Empty<MediaPart>();
    }
}
```

**Comment by sfmskywalker** [comment]
Fixed in 9a2eaf7ac9a4cc96e55fcd73eef2a9f26e49ca43
